theano/tensor/nnet/abstract_conv.py

```diff
@@ -366,6 +374,215 @@ def conv2d_grad_wrt_weights(input,
+    for ii in range(-ratio + 1, ratio):
+        kern[ii + ratio - 1] = ratio - abs(ii)
+    if normalize:
+        kern /= float(ratio)
+    return kern
+
+
+def bilinear_upsampling(input,
+                        ratio,
+                        input_shape=None,
+                        use_1D_kernel=True,
+                        repeat_first_row=True,
+                        repeat_first_col=True):
+    """This function takes a stack of 2D inputs and upsamples it (in 2D space)
+    by the given ratio using bilinear interpolation with fixed (not learned)
+    kernels.
+
+    :type input: symbolic 4D tensor
+    :param input: mini-batch of feature map stacks, of shape
+        (batch size, input channels, input rows, input columns).
+        See the optional parameter ``input_shape`` that will be upsampled.
+
+    :type ratio: integer
+    : param ratio: the ratio by which the input is upsampled
+        in the 2D space.
+
+    :type input_shape: None, tuple/list of len 4 of int or
+        Constant variable.
+    :param input_shape: The shape of the input variable.
+        Optional, possibly used to choose an optimal implementation.
+        You can give ``None`` for any element of the list to specify that this
+        element is not known at compile time.
+
+    :type use_1D_kernel: bool
+    :param use_1D_kernel: if set to true, row and column will
+        be upsampled seperately by 1D kernels, otherwise they are
+        upsampled together using a 2D kernel. The final result is the
+        same, only the speed can differ, given factors such as
+        upsampling ratio.
+
+    :type repeat_first_row: bool
+    :param repeat_first_row: when upsampling by an even ratio, setting
+        this variable to true copies the first row (making it appear twice
+        in the final upsampled tensor), on the other hand a false value
+        copies the last row.
+
+    :type repeat_first_col: bool
+    :param repeat_first_col: when upsampling by an even ratio, setting
+        this variable to true copies the first column (making it appear twice
+        in the final upsampled tensor), on the other hand a false value
+        copies the last column.
+
+    returns: the upsampled 4D tensor, upsampled using bilinear
+             interpolation with fixed (not learned) kernels such that
+             in the 2D space, each feature is upsampled by the given ratio.
+    """
+
+    T = theano.tensor
+    bs, ch, row, col = input.shape
+    up_input = input.reshape((-1, 1, row, col))
+
+    # indicating whether the output dimension
+    # is one dimension less (in the case of even ratio)
+    # or not (in the case of odd ratio)
+    if ratio % 2 == 0:
+        sub = 1
+    else:
+        sub = 0
+
+    # concatenating the first and last row and column
+    # first row
+    concat_mat = T.concatenate((up_input[:, :, 0:1, :], up_input), axis=2)
+    # last row
+    concat_mat = T.concatenate((concat_mat, concat_mat[:, :, -1:, :]), axis=2)
```

Comment:
This is the graph for concat_mat after two concat operations, is it correct?

Join [@A] ''   
 |TensorConstant{2} [@B]
 |Join [@C] ''   
 | |TensorConstant{2} [@B]
 | |Subtensor{::, ::, int64:int64:, ::} [@D] ''   
 | | |Reshape{4} [@E] ''   
 | | | |<TensorType(float64, 4D)> [@F]
 | | | |MakeVector{dtype='int64'} [@G] ''   
 | | |   |Elemwise{Cast{int64}} [@H] ''   
 | | |   | |TensorConstant{-1} [@I]
 | | |   |Elemwise{Cast{int64}} [@J] ''   
 | | |   | |TensorConstant{1} [@K]
 | | |   |Subtensor{int64} [@L] ''   
 | | |   | |Shape [@M] ''   
 | | |   | | |<TensorType(float64, 4D)> [@F]
 | | |   | |Constant{2} [@N]
 | | |   |Subtensor{int64} [@O] ''   
 | | |     |Shape [@M] ''   
 | | |     |Constant{3} [@P]
 | | |Constant{0} [@Q]
 | | |Constant{1} [@R]
 | |Reshape{4} [@E] ''   
 |Subtensor{::, ::, int64::, ::} [@S] ''   
   |Join [@C] ''   
   |Constant{-1} [@T]